### PR TITLE
Expand documentation backends/providers and primitives

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit expands the documentation around primitives and their relationships to backends and providers. Primarily it wasn't very clear from the documentation that the expectation was for the best implementation a provider should implement their own primitives with custom backend specific pre and post processing in order to enable the best quality and most efficient results. It also wasn't clear that the `BackendSampler` and `BackendEstimator` classes were generic and meant to provide compatibility shim between providers without native implementations of the primitive interfaces and don't include any optimizations. This commit expands on these points by improving the class docstrings for `BackendSampler` and `BackendEstimator` and also adding a section to the writing a provider guide on the relationship between primitives and providers.

### Details and comments

### CODE REVIEW

**Review by t-imamichi — DISMISSED** [review]

File: qiskit/providers/__init__.py [hunk 1]
```diff
@@ -612,6 +612,35 @@ def result(self):
...
+the :mod:`qiskit.primitives` module is to explicitly enable this as the
+primitive objects are higher level abstractions to produce processed higher
+level outputs (such as probability distributions and expectation values)
+that abstract away the mechanics of getting the best result efficienctly to
+concentrate on higher level applications using these outputs.
+
+For example, if you're backends were well suited to leverage
+`mthree <https://github.com/Qiskit-Partners/mthree/>`__ measurement
+mitigation to improve the quality of the results you could implement a
+provider specific :class:`~.Sampler` implementation that leverages the
```
Comment (jakelishman) [inline qiskit/providers/__init__.py:634]: `Sampler` or `BaseSampler`?

**Review by jakelishman — COMMENTED** [review]
I put some grammar tweaks in mtreinish/qiskit-core#18, rather than pushing them directly (and iirc github doesn't much like attaching git patch files).  Other than that (and the question below), this looks good to me, and I 100% support improving the documentation.

File: qiskit/providers/__init__.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/providers/__init__.py]: Yeah it should be `BaseSampler` 

```suggestion
provider specific :class:`~.BaseSampler` implementation that leverages the
```

**Review by jakelishman — APPROVED** [review]
Looks good to me now, and since there were no substantive changes since @t-imamichi reviewed, I'll just move this ahead to merge.

**Pull request merged**